File: flutter/wire/lib/src/wire_container.dart
import 'package:get_it/get_it.dart';

/// Dependency injection container backed by [GetIt].
///
/// Feature modules call [register] during their [AppModule.register] phase.
/// Services are later resolved via [resolve].
class WireContainer {
  final GetIt _locator;

  WireContainer({GetIt? locator}) : _locator = locator ?? GetIt.instance;

  /// Register a singleton [instance] of type [T].
  void register<T extends Object>(T instance) {
    if (!_locator.isRegistered<T>()) {
      _locator.registerSingleton<T>(instance);
    }
  }

  /// Register a lazy singleton factory of type [T].
  void registerFactory<T extends Object>(T Function() factory) {
    if (!_locator.isRegistered<T>()) {
      _locator.registerLazySingleton<T>(factory);
    }
  }

  /// Resolve a registered service of type [T].
  /// Throws [StateError] if [T] is not registered.
  T resolve<T extends Object>() {
    assert(_locator.isRegistered<T>(),
        'WireContainer: $T is not registered. Did you call register<$T>() in your AppModule?');
    return _locator<T>();
  }
}

File: ios/Packages/WireCore/Sources/WireCore/WireContainer.swift
import Foundation

/// Lightweight dependency injection container.
///
/// Wraps a dictionary-based service locator.
/// For production use, replace the internals with Swinject.
public final class WireContainer {
    private var services: [ObjectIdentifier: Any] = [:]

    public init() {}

    /// Register a singleton instance of type T.
    public func register<T>(_ instance: T) {
        services[ObjectIdentifier(T.self)] = instance
    }

    /// Resolve a registered service of type T.
    /// Triggers a `fatalError` in debug if T is not registered.
    public func resolve<T>() -> T {
        guard let service = services[ObjectIdentifier(T.self)] as? T else {
            fatalError("WireContainer: \(T.self) is not registered.")
        }
        return service
    }

    /// Resolve a registered service of type T, returning nil if not registered.
    public func tryResolve<T>() -> T? {
        services[ObjectIdentifier(T.self)] as? T
    }
}
